src/core/thread/lowpan.cpp

```diff
@@ -1206,58 +1207,115 @@ int Lowpan::Decompress(Message &           aMessage,
     return (error == OT_ERROR_NONE) ? static_cast<int>(compressedLength) : -1;
 }
 
-otError MeshHeader::Init(const uint8_t *aFrame, uint16_t aFrameLength)
+//---------------------------------------------------------------------------------------------------------------------
+// MeshHeader
+
+void MeshHeader::Init(uint16_t aSource, uint16_t aDestination, uint8_t aHopsLeft)
 {
-    otError error = OT_ERROR_NONE;
+    mSource      = aSource;
+    mDestination = aDestination;
+    mHopsLeft    = aHopsLeft;
+}
+
+bool MeshHeader::IsMeshHeader(const uint8_t *aFrame, uint16_t aFrameLength)
+{
+    return (aFrameLength >= kMinHeaderLength) && ((*aFrame & kDispatchMask) == kDispatch);
+}
+
+otError MeshHeader::ParseFrom(const uint8_t *aFrame, uint16_t aFrameLength, uint16_t &aHeaderLength)
+{
+    otError error = OT_ERROR_PARSE;
+    uint8_t dispatch;
+
+    VerifyOrExit(aFrameLength >= kMinHeaderLength);
+    dispatch = *aFrame++;
 
-    VerifyOrExit(aFrameLength >= 1, error = OT_ERROR_PARSE);
-    mDispatchHopsLeft = *aFrame++;
-    aFrameLength--;
+    VerifyOrExit((dispatch & (kDispatchMask | kSourceShort | kDestShort)) == (kDispatch | kSourceShort | kDestShort));
 
-    if (IsDeepHopsLeftField())
+    mHopsLeft = (dispatch & kHopsLeftMask);
+
+    if (mHopsLeft == kDeepHopsLeft)
     {
-        VerifyOrExit(aFrameLength >= 1, error = OT_ERROR_PARSE);
-        mDeepHopsLeft = *aFrame++;
-        aFrameLength--;
+        mHopsLeft     = *aFrame++;
+        aHeaderLength = kDeepHopsHeaderLength;
     }
     else
     {
-        mDeepHopsLeft = 0;
+        aHeaderLength = kMinHeaderLength;
     }
 
-    VerifyOrExit(aFrameLength >= sizeof(mAddress), error = OT_ERROR_PARSE);
-    memcpy(&mAddress, aFrame, sizeof(mAddress));
+    mSource      = ReadUint16(aFrame);
+    mDestination = ReadUint16(aFrame + 2);
+
+    error = OT_ERROR_NONE;
 
 exit:
     return error;
 }
 
-otError MeshHeader::Init(const Message &aMessage)
+otError MeshHeader::ParseFrom(const Message &aMessage, uint16_t aOffset)
 {
-    otError  error  = OT_ERROR_NONE;
-    uint16_t offset = 0;
-    uint16_t bytesRead;
+    uint16_t headerLength;
+
+    return ParseFrom(aMessage, aOffset, headerLength);
+}
+
+otError MeshHeader::ParseFrom(const Message &aMessage, uint16_t aOffset, uint16_t &aHeaderLength)
+{
+    uint8_t  frame[kDeepHopsLeft];
```

Comment:
Why size the buffer based on `kDeepHopsLeft`?

Reply:
It's typo. Will change to `kDeepHopsHeaderLength`